pennylane/capture/autograph/ag_primitives.py

```diff
@@ -84,6 +94,90 @@ def functional_cond():
     set_state(results)
 
 
+def assert_iteration_inputs(inputs, symbol_names):
+    """All loop carried values, variables that are updated each iteration or accessed after the
+    loop terminates, need to be initialized prior to entering the loop.
+
+    The reason is two-fold:
+      - the type information from those variables is required for tracing
+      - we want to avoid access of a variable that is uninitialized, or uninitialized in a subset
+        of execution paths
+
+    Additionally, these types need to be valid JAX types.
+    """
+
+    for i, inp in enumerate(inputs):
+        if isinstance(inp, Undefined):
+            raise AutoGraphError(
+                f"The variable '{inp}' is potentially uninitialized:\n"
+                " - you may have forgotten to initialize it prior to accessing it inside a loop, or"
+                "\n"
+                " - you may be attempting to access a variable local to the body of a loop in an "
+                "outer scope.\n"
+                f"Please ensure '{inp}' is initialized with a value before entering the loop."
+            )
+
+        try:
+            jax.api_util.shaped_abstractify(inp)
+        except TypeError as e:
+            raise AutoGraphError(
+                f"The variable '{symbol_names[i]}' was initialized with type {type(inp)}, "
+                "which is not compatible with JAX. Typically, this is the case for non-numeric "
+                "values.\n"
+                "You may still use such a variable as a constant inside a loop, but it cannot "
+                "be updated from one iteration to the next, or accessed outside the loop scope "
+                "if it was defined inside of it."
+            ) from e
+
+
+def assert_iteration_results(inputs, outputs, symbol_names):
+    """The results of a for loop should have the identical type as the inputs, since they are
+    "passed" as inputs to the next iteration. A mismatch here may indicate that a loop carried
+    variable was initialized with wrong type.
```

Comment:
```suggestion
    """The results of a for loop should have the identical type as the inputs since they are
    "passed" as inputs to the next iteration. A mismatch here may indicate that a loop-carried
    variable was initialized with the wrong type.
```

Directly from Grammarly : )